### PR TITLE
PyGRB: Update postprocessing functions

### PR DESCRIPTION
**Pull request by jakeb245**
This PR is another step in resolving #4419. It should finish off converting the backend parts of the old XML functionality into HDF. The functionality is intended to be the same, but there could be some cases in here where it should be changed.

### CODE REVIEW

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (pannarale) [inline pycbc/results/pygrb_postprocessing_utils.py:467]: Using `slide_id` to run over `slide_dict` and assign `trigs['network/slide_id'][i]` makes these lines of code hard to parse.

```suggestion
        for i, _ in enumerate(trigs['network/event_id']):
            idx = trigs['network/slide_id'][i]
            sorted_trigs[idx].append(i)
```

**Review by pannarale — DISMISSED** [review]

File: bin/pycbc_multi_inspiral
Comment (jakeb245) [reply bin/pycbc_multi_inspiral:637]: The template ID is stored already. Storing the masses is handy for the chirp mass calculation which happens down the line. The old code stored the template ID as well as the masses. I believe it also wrote the spins for each trigger, which I did not find the need for. This might be easier than adding extra dependencies on the template bank file.

Comment (pannarale) [reply bin/pycbc_multi_inspiral:637]: I don't see the template ID listed in `network_out_vals`.  If we want to avoid the dependency on the template bank, then we should copy all `template.params`.  Otherwise this will be outdated if the template parameters are expanded (e.g., say we have a template bank with tidal deformabilities or precession: then we have to go back and expand this to grab lambda's and spin components).

Comment (jakeb245) [reply bin/pycbc_multi_inspiral:637]: The template ID is written elsewhere (looks like it's done [here](https://github.com/gwastro/pycbc/blob/master/pycbc/events/eventmgr.py#L653)). I've confirmed it shows up in the output files as well.

**Comment by titodalcanton** [comment]
@jakeb245 yes, I would use a similar approach used by `pycbc_inspiral` in the all-sky all-time search, which indeed consists in storing just the template hash for each trigger after matched filtering. To be precise, the following happens there:

1. `pycbc_inspiral` does matched filtering for each bank split separately, and the corresponding triggers store the template hashes only.
2. `pycbc_coinc_mergetrigs` then takes all the triggers from all the bank splits from step 1, merges them together, and uses the template hashes to recover the template IDs into the full bank.

Have a look at this documentation page for more info: https://pycbc.org/pycbc/latest/html/formats/hdf_format.html. You can also see an example of this by looking at the various files from one of the all-sky all-time O4 chunks, in particular you want to look at one of the `full_data/*-INSPIRAL_BANK*.hdf` files for step 1, and `full_data/*-HDF_TRIGGER_MERGE_FULL_DATA*.hdf` for step 2.

Ping me on Slack if more is needed.

File: bin/pygrb/pycbc_grb_inj_finder
Comment (titodalcanton) [inline bin/pygrb/pycbc_grb_inj_finder:340]: Should this not be `if col == "template_id":`? Or does `col` contain other characters in addition?

File: bin/pygrb/pycbc_grb_trig_combiner
Comment (titodalcanton) [inline bin/pygrb/pycbc_grb_trig_combiner:215]: Same comment as before, should this not be an exact match?

File: bin/pygrb/pycbc_make_offline_grb_workflow
Comment (titodalcanton) [inline bin/pygrb/pycbc_make_offline_grb_workflow:414]: Is the first item of `bank_files` always guaranteed to be the full bank? Or is `bank_files` always going to contain a single item? This is not obvious from the code. A comment would be useful, or maybe a more self-explanatory choice of variables? Do not bother if too complicated though.

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (titodalcanton) [inline pycbc/results/pygrb_postprocessing_utils.py:467]: Could this be instead
```python
for i in range(len(trigs['network/event_id'])):
```

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (titodalcanton) [inline pycbc/results/pygrb_postprocessing_utils.py:522]: Can you confirm that the BestNR calculation happens somewhere else (earlier) in the new workflow? I.e. if I wanted to change `opts.newsnr_threshold`, where would that be applied?

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (titodalcanton) [inline pycbc/results/pygrb_postprocessing_utils.py:801]:
```suggestion
    with h5py.File(bank_path, "r") as bank:
        hashes = bank['template_hash'][:]
```
so you close the bank file as soon as you are done with it.

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (titodalcanton) [inline pycbc/results/pygrb_postprocessing_utils.py:803]: Why is the `template_hash` dataset stored inside the ifo-specific groups, instead of the `network` group? Is that not unnecessarily duplicating the same dataset for each ifo?

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (titodalcanton) [reply pycbc/results/pygrb_postprocessing_utils.py:467]: I am a bit puzzled by this code though. Imagine that `trigs['network/event_id']` contains [2,3,0,1]. This would mean that `i` would go through the list [0,1,2,3] instead. Is that the correct thing to do? Somewhat confusing that this code does not appear to actually use the values in `event_id`.

**Comment by titodalcanton** [comment]
@jakeb245 I had a close look at this PR. As you know I am a bit removed from the development of this code, so I left a bunch of questions above to fill some gaps in my understanding, but I do not see anything too concerning here. I am happy to approve this once you have answered the comments.

File: bin/pygrb/pycbc_grb_inj_finder
Comment (jakeb245) [reply bin/pygrb/pycbc_grb_inj_finder:340]: `col` is the full column header, so it includes the IFO/network tags.

File: bin/pygrb/pycbc_make_offline_grb_workflow
Comment (jakeb245) [reply bin/pygrb/pycbc_make_offline_grb_workflow:414]: Good point. If we were to run with two bank files this would become an issue. If there's only one bank then the `bank_files` object should just the single file. I suppose it would be worth having the workflow generator allow for multiple bank files. I don't think it would be complicated to do that, but that case would cause problems with the template hash/ID mapping.

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (jakeb245) [reply pycbc/results/pygrb_postprocessing_utils.py:522]: Yes, currently the reweighted SNR is calculated in `pycbc_multi_inspiral` at around [here](https://github.com/gwastro/pycbc/blob/e2c35eff8367fa4700588a60f80737949d8fd254/bin/pycbc_multi_inspiral#L613). The `newsnr_threshold` option is also set in that executable.

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (jakeb245) [reply pycbc/results/pygrb_postprocessing_utils.py:803]: I'm not sure. The writing to file is done [here](https://github.com/gwastro/pycbc/blob/e2c35eff8367fa4700588a60f80737949d8fd254/pycbc/events/eventmgr.py#L787). The line that makes me unsure is [the definition of the index](https://github.com/gwastro/pycbc/blob/e2c35eff8367fa4700588a60f80737949d8fd254/pycbc/events/eventmgr.py#L719). It seems dependent on the ifo, but one would expect them to be the same.

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (jakeb245) [reply pycbc/results/pygrb_postprocessing_utils.py:467]: I think working with direct indices would be faster rather than having to do something like `trigs['network/reweighted_snr']['network/event_id' == event_id]`. I'm thinking it would be more intuitive to have this function output the `event_id` though, so in this case I think I'll change that.

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (titodalcanton) [reply pycbc/results/pygrb_postprocessing_utils.py:806]: Yeah, what I see in `eventmgr.py` does not look correct, or at least not efficient. All template properties should be identical across interferometers. Maybe we do not need to worry about this in this PR, but it should be addressed shortly after this is merged. Please bring this up at the development call.

File: bin/pygrb/pycbc_grb_inj_finder
Comment (titodalcanton) [reply bin/pygrb/pycbc_grb_inj_finder:340]: Are you sure? In the code above I see
```python
for x in triggers:
    ij = x.split('/')
    out_trigs[ij[0]][ij[1]].append(triggers[x][eid])
```
so it seems to me that `out_trigs` is a dict-of-dict type object, where the outer key is the interferometer or "network", and the inner key is the field name.

File: bin/pygrb/pycbc_make_offline_grb_workflow
Comment (titodalcanton) [reply bin/pygrb/pycbc_make_offline_grb_workflow]: I do not see what use case we might want to support where there are two banks in a single workflow. Not in the immediate future, at least.

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (titodalcanton) [reply pycbc/results/pygrb_postprocessing_utils.py]: Not sure I follow. Presumably the code using the actual `event_id`s would be as simple as
```python
for i in trigs['network/event_id']:
```
Is that not the case?

File: bin/pygrb/pycbc_make_offline_grb_workflow
Comment (titodalcanton) [reply bin/pygrb/pycbc_make_offline_grb_workflow]: I would do the "Multiple banks not supported" check early on, immediately after the definition of `bank_files` in `pycbc_make_offline_grb_workflow`. Then define `full_bank_file = bank_files[0]`. In that way you make it obvious from the code that there is a single full bank file.

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (jakeb245) [reply pycbc/results/pygrb_postprocessing_utils.py]: Good point. The only time the index gets used is to pull the `slide_id`. This operation is likely faster, but probably not worth using `enumerate` for.

File: bin/pygrb/pycbc_grb_inj_finder
Comment (jakeb245) [reply bin/pygrb/pycbc_grb_inj_finder:340]: Oh right, I was thinking about `pycbc_grb_trig_combiner`. You're correct!

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (titodalcanton) [reply pycbc/results/pygrb_postprocessing_utils.py:807]: I looked a bit more at the `eventmrg.py` code and the situation is more confusing than I thought, so I opened https://github.com/gwastro/pycbc/issues/4632 to track that issue.

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (titodalcanton) [reply pycbc/results/pygrb_postprocessing_utils.py]: Ah, I think I understand the goal now. Each trigger has an entry in the `event_id` and `slide_id` datasets (at the same index in the two) and you want to group the triggers by `slide_id`. So what you have now is probably very slow, but how about this:
```python
    for slide_id, event_id in zip(trigs['network/slide_id'], trigs['network/event_id']):
        sorted_trigs[slide_id].append(event_id)
```
Also, note that in a couple places you do `'network/event_id' == event_id` which does not make sense, you probably meant `trigs['network/event_id'] == event_id`.

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (titodalcanton) [inline pycbc/results/pygrb_postprocessing_utils.py:478]: Not sure this will work, I think what you need is
```suggestion
            index = numpy.flatnonzero(trigs['network/event_id'] == event_id)[0]
            end_time = trigs['network/end_time_gc'][index]
```

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (titodalcanton) [inline pycbc/results/pygrb_postprocessing_utils.py:496]: Same as above, if I understand the involved data structures correctly, this should produce an error. This should be what you want:
```suggestion
                                  if trigs['network/end_time_gc'][
                                      trigs['network/event_id'] == event_id][0]
```
The idea is that `trigs['network/event_id'] == event_id` will give you a boolean array of the same shape as `trigs['network/event_id']`, with a single True at the position of the right event. You then use that array to index `trigs['network/end_time_gc']`, which will give you a single-entry array of the end time. Then `[0]` picks out the only entry as a scalar.

**Comment by titodalcanton** [comment]
I think there are a couple things to fix (noted above), then I think this can go in.

File: pycbc/results/pygrb_postprocessing_utils.py
Comment (jakeb245) [reply pycbc/results/pygrb_postprocessing_utils.py:496]: This is definitely what I was intending to do. Thanks for catching that!

**Review by titodalcanton — APPROVED** [review]

**Comment by titodalcanton** [comment]
I am still not super convinced that I understand the logic of the slide_id/event_id thing here, so whoever is developing that functionality should keep it in mind. But let's merge this and continue developing.

**Pull request merged**
